src/zarr/core/metadata/dtype.py

```diff
@@ -0,0 +1,806 @@
+
+
+def structured_scalar_to_json(data: bytes, zarr_format: ZarrFormat) -> str:
+    if zarr_format == 2:
+        return base64.b64encode(data).decode("ascii")
+    raise NotImplementedError(f"Invalid zarr format: {zarr_format}. Expected 2.")
+
+
+def structured_scalar_from_json(data: str, zarr_format: ZarrFormat) -> bytes:
+    if zarr_format == 2:
+        return base64.b64decode(data.encode("ascii"))
+    raise NotImplementedError(f"Invalid zarr format: {zarr_format}. Expected 2.")
+
+
+def float_from_json_v2(data: JSONFloat) -> float:
+    match data:
+        case "NaN":
+            return float("nan")
+        case "Infinity":
+            return float("inf")
+        case "-Infinity":
+            return float("-inf")
+        case _:
+            return float(data)
+
+
+def float_from_json_v3(data: JSONFloat) -> float:
+    # todo: support the v3-specific NaN handling
+    return float_from_json_v2(data)
+
+
+def float_from_json(data: JSONFloat, zarr_format: ZarrFormat) -> float:
+    if zarr_format == 2:
+        return float_from_json_v2(data)
+    else:
+        return float_from_json_v3(data)
+
+
+def complex_from_json_v2(data: JSONFloat, dtype: Any) -> np.complexfloating[Any, Any]:
+    return dtype.type(complex(*data))
+
+
+def complex_from_json_v3(
+    data: tuple[JSONFloat, JSONFloat], dtype: Any
+) -> np.complexfloating[Any, Any]:
+    return dtype.type(complex(*data))
+
+
+def complex_from_json(
+    data: tuple[JSONFloat, JSONFloat], dtype: Any, zarr_format: ZarrFormat
+) -> np.complexfloating:
+    if zarr_format == 2:
+        return complex_from_json_v2(data, dtype)
+    else:
+        if check_json_complex_float_v3(data):
+            return complex_from_json_v3(data, dtype)
+        else:
+            raise TypeError(f"Invalid type: {data}. Expected a sequence of two numbers.")
+    raise ValueError(f"Invalid zarr format: {zarr_format}. Expected 2 or 3.")
+
+
+def datetime_to_json(data: np.datetime64[Any]) -> int:
+    return data.view(np.int64).item()
+
+
+def datetime_from_json(data: int, unit: DateUnit | TimeUnit) -> np.datetime64[Any]:
+    return np.int64(data).view(f"datetime64[{unit}]")
+
+
+TScalar = TypeVar("TScalar", bound=np.generic | str, covariant=True)
+# TODO: figure out an interface or protocol that non-numpy dtypes can
+TDType = TypeVar("TDType", bound=np.dtype[Any])
+
+
+@dataclass(frozen=True, kw_only=True)
+class DTypeWrapper(Generic[TDType, TScalar], ABC, Metadata):
+    name: ClassVar[str]
+    dtype_cls: ClassVar[type[TDType]]  # this class will create a numpy dtype
+    endianness: Endianness | None = "native"
+
+    def __init_subclass__(cls) -> None:
+        # TODO: wrap this in some *very informative* error handling
+        generic_args = get_args(get_original_bases(cls)[0])
+        # the logic here is that if a subclass was created with generic type parameters
+        # specified explicitly, then we bind that type parameter to the dtype_cls attribute
```

Comment:
I've read this like 4 times and still don't feel like I understand what it actually does, or understand where this logic gets used by any of the subclasses below.

Reply:
i removed the use of `__init_subclass__`. What I was using it for was ensuring that the following pattern would work:

```
class Base(Generic[T])
  t_cls: type(T)
  
  make_t(self) -> T
    ...

  def __init_subclass__(cls):
    # bind the generic type parameter to the t_cls class attribute
    ...

class ToWrap:
  ...

X(Base[ToWrap[):
  ...

assert X.t_cls == ToWrap
assert X.make_t() == ToWrap()
```
I.e., binding the generic type parameter in a subclass to a particular attribute (`dtype_cls`).
But I can do the same thing more directly by just binding that attribute in the class definition, without weird generic type magic.